### ISSUE TITLE
MCP-0025: Functions with combined input output arguments

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/ModelicaSpecification#2012]
**Reported by rfranke on 15 Jun 2016 19:21 UTC**
Modelica functions must not write to inputs. This creates the need for unnecessary copy operations, which is costly for arrays. This proposal is about extending Modelica functions to supporting combined input output arguments.

See:
https://svn.modelica.org/projects/MCP/public/MCP-0025_InputOutput/MCP-0025_InputOutput.docx

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/2012

**Comment by modelica-trac-importer** [comment]
**Comment by rfranke on 16 Jun 2016 05:26 UTC**
r9376 enhances the proposal such that it becomes backwards compatible.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 16 Jun 2016 08:34 UTC**
The MCP doesn't explain why this is necessary.
A compiler already has the information to perform this (as will be explained) - and the example isn't convincing: copy is not ultra-cheap - but cheap compared to the rest of dgesv.

There are cases where copying is expensive - but that is much more specialized cases. (See 3.3 in https://modelica.org/events/modelica2015/proceedings/html/submissions/ecp15118235_ElmqvistOlssonGotemanRoxlingZimmerPollok.pdf )

For the work-area a compiler can perform the optimization without any new syntax: the compiler can see that only one of Awork and A is sent to the external function, and also see if A is a temporary. Thus if A is a temporary there is no need to copy it once more, and if not a temporary we have to copy it anyway.

The same logic of handling this automatically without any new language constructs can be applied to the output-handling, with two caveats, and unfortunately the MCP does not discuss them.
The first issue for output-handling is aliasing if calling `LAPACK.dgesv(output A, output A)` - and similar variants.
The second issue is error handling. "Commit or rollback semantics" is often a good idea for safety, but we are now breaking that.

**Comment by modelica-trac-importer** [comment]
**Comment by rfranke on 16 Jun 2016 20:10 UTC**
> A compiler already has the information to perform this

This sounds tricky to implement. Are you aware of any Modelica compiler that analyses the function body to optimize a function call?

The point with this MCP is that the optimization can be done  basing on a function signature, which is checked during translation anyway.

> The first issue for output-handling is aliasing

This can be solved for sure, e.g. disallowed.

> The second issue is error handling.

Does Modelica have error handling (beyond raising the error and terminating the simulation)?

**Comment by modelica-trac-importer** [comment]
**Comment by rfranke on 17 Jun 2016 06:05 UTC**
r9382 extends the summary of MCP-0025 to provide a better motivation of combined input/output arguments.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 17 Jun 2016 08:16 UTC**
Replying to [comment:3 rfranke]:
> > A compiler already has the information to perform this
> 
> This sounds tricky to implement. Are you aware of any Modelica compiler that analyses the function body to optimize a function call?

Well, Dymola (and 3D Experience Platform) analyzes the contents of the function call in various ways - some quite advanced, even if we by design don't always use them. And this specific analysis seems trivial to implement to me, and with minimal cost, even if not yet implemented.

It is basically just a matter of seeing "Awork=A", and noticing that "Awork" but not "A" is sent to the external function, so we can skip that copy - if Awork is a temporary. The analysis of the calling site is similarly simple (except for the aliasing).

That is not a complicated analysis, and requiring a keyword 'output' to perform the analysis seems excessive - and I am in general against this idea of changing the language to enforce a specific optimization that tools could perform even without the change.

The reasons are both that it distorts the priorities for tool developers - there are many potential optimizations and we generally implement them based on some idea of sensible priorities, whereas this bumps a particular optimization to the front of the line, and that it adds complexity to the specification - which unnecessarily increases the barrier of entry.

And also because it forces library developers to add complexity to their libraries, making the libraries more complicated to understand for users.

> > The first issue for output-handling is aliasing
> 
> This can be solved for sure, e.g. disallowed.

Yes, but aliasing requires a complicated analysis to detect.
A lot more complicated than the ones needed for the optimization without aliasing.

> > The second issue is error handling.
> 
> Does Modelica have error handling (beyond raising the error and terminating the simulation)?

Modelica tools have more advanced error handling than terminating the simulation when an error is detected. 

That is allowed in the MSL specification by section 8.3.7 (and 12.9.6 that states that ModelicaError works similarly as an assert in the code):
"The simulation may continue with another evaluation [e.g., with a shorter step-size, or by changing the values of iteration variables]. If the simulation is aborted, message indicates the cause of the error."

However, even if this is a potential problem it seems that even if a failed evaluation may mess up some values I don't see how that will cause problems for the next evaluation (because the iteration variables are likely not sent to the function). But that requires a more detailed analysis.

**Comment by modelica-trac-importer** [comment]
**Comment by rfranke on 17 Jun 2016 09:17 UTC**
> The reasons are both that it distorts the priorities for tool developers - there are many potential optimizations and we generally implement them based on some idea of sensible priorities, whereas this bumps a particular optimization to the front of the line, 

Well, you should not mess up development plans with language specification.

> and that it adds complexity to the specification - which unnecessarily increases the barrier of entry.

It is a backwards compatible extension. Nothing changes if people don't use this feature. Moreover, anyone who wants to write textual Modelica functions should have some basic programming knowledge. He/she might wonder why combined input/output known from other languages, like Python or Julia, besides C and F77 and many more, does not find in Modelica.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 17 Jun 2016 11:50 UTC**
Replying to [comment:6 rfranke]:
> > The reasons are both that it distorts the priorities for tool developers - there are many potential optimizations and we generally implement them based on some idea of sensible priorities, whereas this bumps a particular optimization to the front of the line, 
> 
> Well, you should not mess up development plans with language specification.

Exactly - and this MCP is doing that, it should be part of the development plans for various tools - not part of the language specification.

The reason is that the copy-avoiding analysis is trivial and can be done with or without changes in the language specification.

An analysis of the function itself takes perhaps an hour to implement - most of the time for all the issues that aren't considered the MCP (the part that the MCP supposedly helps with takes considerably less than that - and I'm certain that it doesn't take more time).

I would even say that the MCP is **counter-productive**, because the changes of the grammar (=actual grammar, pretty-printing, GUI; they are all also fairly straightforward and trivial) will for some us take more time to implement than the copy-avoiding-analysis that they are supposed to help with.

> > and that it adds complexity to the specification - which unnecessarily increases the barrier of entry.
> 
> It is a backwards compatible extension. Nothing changes if people don't use this feature. 

Doesn't matter. It adds to the complexity regardless of that - and someone looking at the language specification will not easily see exactly which features are needed and which ones aren't.

**Comment by modelica-trac-importer** [comment]
**Comment by rfranke on 17 Jun 2016 22:17 UTC**
It might really be a good idea to conserve the current status of Modelica and focus efforts on something new.

**Comment by modelica-trac-importer** [comment]
**Comment by jmattsson on 20 Jun 2016 07:29 UTC**
Replying to [comment:6 rfranke]:
> It is a backwards compatible extension. Nothing changes if people don't use this feature. Moreover, anyone who wants to write textual Modelica functions should have some basic programming knowledge. He/she might wonder why combined input/output known from other languages, like Python or Julia, besides C and F77 and many more, does not find in Modelica.

I disagree - most languages do not have combined input/output arguments. They do have pass by reference, which enables you to change variables in the calling scope, which in turn enables you to emulate combined input/output arguments _at your own risk_. Neither combined input/output arguments or passing non-constant data by reference goes well with mathematical functions.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 20 Jun 2016 14:04 UTC**
Discussion:
J. Mattsson:
Moving away from Mathematical functions (not really pure functions anymore).
Need rule for forbidding "input output" components in models, etc.
C. Höger:
Standard ML - annotate fields to say if mutable: add complication of garbage-collection due to loss of referential transparency. Not clear if worth the effort.
H. Olsson:
See no need to change language to allow this optimization. Due to support of both C and Fortran a matrix will be copied in some cases (since it is transposed), so it will not work exactly as a reference.

One possibility would be to suggest this optimization for external functions (together with stating that Awork is needed).
\-
Should we proceed with this MCP? Conclusion seems to be to put in on hold - and combine with higher-level data-types.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 23 Jun 2016 11:54 UTC**
The conclusion was to put this on hold waiting on higher-level data-types.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 28 Feb 2017 12:38 UTC**
To make it easy to get overview these tickets that are effectively on hold are closed as expired. They can be re-opened later if/when we start working on them.
